speechbrain/lobes/models/transformer/TransformerUtils.py

```diff
@@ -0,0 +1,621 @@
+            The mask for the tgt keys per batch (optional).
+        memory_key_padding_mask : tensor
+            The mask for the memory keys per batch (optional).
+        """
+        output = tgt
+        self_attns, multihead_attns = [], []
+        for dec_layer in self.layers:
+            output, self_attn, multihead_attn = dec_layer(
+                output,
+                memory,
+                tgt_mask=tgt_mask,
+                memory_mask=memory_mask,
+                tgt_key_padding_mask=tgt_key_padding_mask,
+                memory_key_padding_mask=memory_key_padding_mask,
+                pos_embs_tgt=pos_embs_tgt,
+                pos_embs_src=pos_embs_src,
+            )
+            self_attns.append(self_attn)
+            multihead_attns.append(multihead_attn)
+        output = self.norm(output)
+
+        return output, self_attns, multihead_attns
+
+
+class NormalizedEmbedding(nn.Module):
+    """This class implements the normalized embedding layer for the transformer.
+
+    Since the dot product of the self-attention is always normalized by sqrt(d_model)
+    and the final linear projection for prediction shares weight with the embedding layer,
+    we multiply the output of the embedding by sqrt(d_model).
+
+    Arguments
+    ---------
+    d_model: int
+        The number of expected features in the encoder/decoder inputs (default=512).
+    vocab: int
+        The vocab size.
+
+    Example
+    -------
+    >>> emb = NormalizedEmbedding(512, 1000)
+    >>> trg = torch.randint(0, 999, (8, 50))
+    >>> emb_fea = emb(trg)
+    """
+
+    def __init__(self, d_model, vocab):
+        super().__init__()
+        self.emb = sb.nnet.embedding.Embedding(
+            num_embeddings=vocab, embedding_dim=d_model, blank_id=0
+        )
+        self.d_model = d_model
+
+    def forward(self, x):
+        return self.emb(x) * math.sqrt(self.d_model)
+
+
+def get_key_padding_mask(padded_input, pad_idx):
+    """Creates a binary mask to prevent attention to padded locations.
+
+    Arguments
+    ----------
+    padded_input: int
+        Padded input.
+    pad_idx:
+        idx for padding element.
+
+    Example
+    -------
+    >>> a = torch.LongTensor([[1,1,0], [2,3,0], [4,5,0]])
+    >>> get_key_padding_mask(a, pad_idx=0)
+    tensor([[False, False,  True],
+            [False, False,  True],
+            [False, False,  True]])
+    """
+    if len(padded_input.shape) == 4:
+        bz, time, ch1, ch2 = padded_input.shape
+        padded_input = padded_input.reshape(bz, time, ch1 * ch2)
+
+    key_padded_mask = padded_input.eq(pad_idx).to(padded_input.device)
+
+    # if the input is more than 2d, mask the locations where they are silence
+    # across all channels
+    if len(padded_input.shape) > 2:
+        key_padded_mask = key_padded_mask.float().prod(dim=-1).bool()
+        return key_padded_mask.detach()
+
+    return key_padded_mask.detach()
+
+
+def get_lookahead_mask(padded_input):
```

Comment:
There are implementations from PyTorch (https://github.com/pytorch/pytorch/blob/9573e7a644544fbd623e652766d47c3e84d588de/torch/nn/modules/transformer.py#L148) and they are probably more efficient?

Is there a reason to reinvent the wheels?

Reply:
@csukuangfj This was already part of SB, I just moved it to clean the code a bit and remove function/classes that are utils instead of being in the TransformerInterface module